src/ape_pm/dependency.py

```diff
@@ -0,0 +1,321 @@
+import json
+import os
+import shutil
+from collections.abc import Iterable
+from functools import cached_property
+from pathlib import Path
+from typing import Optional, Union
+
+from pydantic import model_validator
+
+from ape.api.projects import DependencyAPI
+from ape.exceptions import ProjectError
+from ape.logging import logger
+from ape.utils._github import github_client
+
+
+class LocalDependency(DependencyAPI):
+    """
+    A dependency located on the local machine.
+    """
+
+    local: Path
+    """
+    The root path (and API defining key) to the dependency files.
+    """
+
+    version: Optional[str] = None
+    """
+    Specified version.
+    """
+
+    @model_validator(mode="before")
+    @classmethod
+    def validate_local_path(cls, model):
+        # Resolves the relative path so if the dependency API
+        # data moves, it will still work.
+        path = Path(model["local"])
+        if path.is_absolute():
+            return model
+
+        elif "project" in model:
+            # Just in case relative paths didn't get resolved.
+            # Note: Generally, they should be resolved at model
+            #  construction time, if parsing a config file normally.
+            project = model.pop("project")
+            model["local"] = (project / path).resolve()
+
+        return model
+
+    @property
+    def package_id(self) -> str:
+        return self.local.as_posix()
+
+    @property
+    def version_id(self) -> str:
+        return self.version or "local"
+
+    @property
+    def uri(self) -> str:
+        return self.local.as_uri()
+
+    def fetch(self, destination: Path):
+        if destination.is_dir():
+            destination = destination / self.name
+
+        if self.local.is_dir():
+            project = self.Project(self.local, config_override=self.config_override)
+            project.unpack(destination)
+        elif self.local.is_file() and self.local.suffix == ".json":
+            # Using a manifest directly as a dependency.
+            if not destination.suffix:
+                destination = destination / self.local.name
+
+            destination.unlink(missing_ok=True)
+            destination.parent.mkdir(parents=True, exist_ok=True)
+            destination.write_text(self.local.read_text())
+
+
+class GithubDependency(DependencyAPI):
+    """
+    A dependency from Github. Use the ``github`` key in your ``dependencies:``
+    section of your ``ape-config.yaml`` file to declare a dependency from GitHub.
+
+    Config example::
```

Comment:
does this do anything special? how does this example render?

Reply:
Two colons mean custom header, you can put whatever you want for the title. We do this for our `Usage example::` everywhere as well. Here is the render (on a laptop web browser):

<img width="745" alt="Screenshot 2024-05-22 at 08 40 10" src="https://github.com/ApeWorX/ape/assets/19540978/37483908-4b7c-4f2e-9a3d-954ede284a67">